### PR TITLE
THC TDA pbc mGW

### PR DESCRIPTION
**Pull request by mkakcl**
Added functionality to complete TDA pbc calculations using imported THC integrals. Structurally very similar to the initial THC implementation for TDA for molecules. Includes the main run file (pbc/thc.py) along with some tests (tests/test_kthcgw.py). Changes are also made in other files to align the method for running the new method with the molecular code.

Additionally, added a feature to allow loading THC integrals into the CDERI methods so that easy comparisons can be made between methods.

No example was created as the functionality is the same as the original THC example, except using the functions with K as per example 06_solids

### CODE REVIEW

**Comment by ghb24** [comment]
I've taken a quick look, and looks good to me - nice work - but I'll leave it to Ollie to review if he wants before merging.
One quick point though (I might have mentioned this before when you implemented thc for molecules if I remember) - it does seem a little weird specifying that you want to use thc via the 'polarizability' keyword, which is really separate to the form of the interaction kernel. Not super important, but might be worth thinking about splitting that to a separate specifier - it's independent to the choice of how the polarizability is built after all...

**Comment by obackhouse** [comment]
> I've taken a quick look, and looks good to me - nice work - but I'll leave it to Ollie to review if he wants before merging. One quick point though (I might have mentioned this before when you implemented thc for molecules if I remember) - it does seem a little weird specifying that you want to use thc via the 'polarizability' keyword, which is really separate to the form of the interaction kernel. Not super important, but might be worth thinking about splitting that to a separate specifier - it's independent to the choice of how the polarizability is built after all...

yeah maybe in theory it could be a different keyword, but the reason it is specified as part of the polarizability is that implementationally, only the polarizability class needs to change between CD and THC implementations.

File: tests/test_kthcgw.py
Comment (obackhouse) [inline tests/test_kthcgw.py:188]: missing newline

**Review by obackhouse — COMMENTED** [review]
Added a few comments, just did a skim read so let me poke around a bit more when I get some more time and i'll review

File: momentGW/base.py
Comment (mkakcl) [reply momentGW/base.py:142]: Currently, functionally they are the same but I deliberately split it off from the `GW.thc_opts["file_path"]` just in case we want to have a method for importing ERIs into a calculation. Would just require an if statement depending on say the file name. Could be useful if you have a long calculation that may fail and don't want to do run the DFT calc. again? though this would require a few more things in the import.

Comment (obackhouse) [reply momentGW/base.py:142]: For CDERIs I would probably prefer this to be handled by the mean field on the PySCF side -- they have functionality to do that. I think it makes more sense for them to be loaded on the PySCF side than for us to add another keyword argument here. Thoughts?

I think that similarly, the file path for the THC integrals should go once there is a proper interface to their generation, but we can't afford that luxury yet obviously.

**Comment by obackhouse** [comment]
as per some of my comments I would rather keep file inputs and file paths to a bare minimum -- only where there is no other obvious solution (i.e. the THC in the current state). I think they make for difficult to maintain code and it makes more sense for us to just include examples showing how one would go about hacking/duck typing loaded integrals in on the PySCF side

File: momentGW/pbc/ints.py
Comment (obackhouse) [inline momentGW/pbc/ints.py:321]: clean up this docstring pls

File: momentGW/pbc/ints.py
Comment (obackhouse) [inline momentGW/pbc/ints.py:322]: newline after the docstring, unless function is trivial

File: momentGW/base.py
Comment (mkakcl) [reply momentGW/base.py:142]: I'm happy to take that direction with it. I just figured given I was putting in something that would import ERIs into the CDERI code, may as well make it general and have this conversation. But happy to use pyscf for all CDERI imports.

As you say, hopefully we can remove all import file paths with a proper THC integral integration, but will have to see.

File: momentGW/pbc/ints.py
Comment (obackhouse) [inline momentGW/pbc/ints.py:326]: `if "thc" not in self.input_path.lower()`

File: momentGW/pbc/ints.py
Comment (obackhouse) [inline momentGW/pbc/ints.py:335]: let's add an `__eq__` to the `momentGW.pbc.kpts.KPoints` class to handle this entire block - the `wrap_around` should handle the sign too. I can do this

File: momentGW/pbc/ints.py
Comment (obackhouse) [inline momentGW/pbc/ints.py:343]: Don't assign new variables in functions that aren't initialised in `__init__` -- add `self._naux = None` to the `__init__` (the one for this class, not a subclass, I'd say), then in the `naux` property do:
```
@property
def naux(self):
  if self._rot is None:
    if self._naux is not None:
      return self._naux
...
```

File: momentGW/pbc/ints.py
Comment (obackhouse) [inline momentGW/pbc/ints.py:357]: `lib.einsum`

File: momentGW/pbc/ints.py
Comment (obackhouse) [inline momentGW/pbc/ints.py:345]: is this tested for MPI? if not, remove any MPI stuff so it runs serially even with `mpirun`

File: momentGW/pbc/ints.py
Comment (obackhouse) [inline momentGW/pbc/ints.py:352]: just to check, these are `h5py` datasets right? I would maybe prefer `np.asarray` just stylistically. `np.asarray` says "copy me if i'm not already a numpy array" whereas `np.array` says "copy me into a new numpy array"

File: momentGW/pbc/ints.py
Comment (obackhouse) [inline momentGW/pbc/ints.py:374]: `lib.einsum`

File: momentGW/pbc/ints.py
Comment (obackhouse) [reply momentGW/pbc/ints.py:654]: see my comments on the `self._naux` initialisation above

File: momentGW/pbc/ints.py
Comment (mkakcl) [reply momentGW/pbc/ints.py]: Removed all references to MPI in relation to thc-tda-k-mgw since it is untested

File: momentGW/pbc/ints.py
Comment (mkakcl) [reply momentGW/pbc/ints.py:342]: Changed this in the original thc implementation so that the new and old method are the same.

**Comment by mkakcl** [comment]
> as per some of my comments I would rather keep file inputs and file paths to a bare minimum -- only where there is no other obvious solution (i.e. the THC in the current state). I think they make for difficult to maintain code and it makes more sense for us to just include examples showing how one would go about hacking/duck typing loaded integrals in on the PySCF side

Based on this conversation, I changed the code so that there is now only a singular import, `thc_opts["file_path"]` which is used for  doing both the THC code and for building CDERIs from THC integrals.

**Comment by mkakcl** [comment]
A more general comment/query, how general/user friendly do we want the THC import functionality to be for the time being? For example the THC import files are design for a specific data structure given what we were sent. In the time between receiving different THC integrals, their data structure changes and so the import process needed to change/didn't work for the old format. So how much guidance should we give people if they have their own? or should it be more there just as an internal tool functionality for the time being?

File: tests/test_kthcgw.py
Comment (mkakcl) [inline tests/test_kthcgw.py:1]: With the `gw.optimise_chempot = True` for `test_nelec`, I lowered the precision from 8 on the molecular to 1 on this solids test. I did this as the precision only matched to this degree for both THC and CDERI methods for solids. I am not very familiar with the `optimise_chempot` method, do we expect it to be as exact for `nelec` in solids as it was for molecules?

File: momentGW/pbc/ints.py
Comment (mkakcl) [reply momentGW/pbc/ints.py:334]: I haven't changed this, but if you don't have time to do it, more than happy to do it myself.

File: tests/test_kthcgw.py
Comment (obackhouse) [reply tests/test_kthcgw.py:1]: Hmm, there could be a scaling issue, but I would expect them to match for THC and CDERI. `optimise_chempot` takes an objective function that can be thought of (pseudo-code) as:
```
def f(x):
  se.energies -= x
  gf = diagonalise(se)
  return (trace(rdm1(gf)) - nelec_true)**2
```
and then, along with the gradient, optimises this shift in the auxiliaries such that the resulting GF can fill its QMOs (c.f. Aufbau principle) in a way that faithfully conserves the number of physical electrons in the original system.

**Comment by obackhouse** [comment]
> A more general comment/query, how general/user friendly do we want the THC import functionality to be for the time being? For example the THC import files are design for a specific data structure given what we were sent. In the time between receiving different THC integrals, their data structure changes and so the import process needed to change/didn't work for the old format. So how much guidance should we give people if they have their own? or should it be more there just as an internal tool functionality for the time being?

I think if their interface is that unstable, then we don't worry about it for the time being -- it would maybe be better suited to a dev branch until we have a stable interface, but if the code is clean and implemented to production quality then it's fine to have it in `master` and just update the interface whenever they do.

**Comment by obackhouse** [comment]
What features, if any, are not supported by this PR? Does it support all self-consistencies, Fock loops etc.?

**Comment by mkakcl** [comment]
> What features, if any, are not supported by this PR? Does it support all self-consistencies, Fock loops etc.?

So I have tested everything seen in an example; evGW (neither, g0, w0), scGW (g0, w0, fock_loop and all combinations) and qsGW (eta, srg and solver (with combinations)). For chempot, see above and there is no UHF implemented. Is there anything else?

File: tests/test_kthcgw.py
Comment (mkakcl) [reply tests/test_kthcgw.py:1]: So the behaviour does match for THC and for CDERIs, however from what I can tell, while the code for `minimize_chempot` is running, it is not having any effect on the number of occupied  in the system. This differs from the molecular case and as described above, where the number becomes much closer to the nelec from the cell (order 10^-8 of precision).

**Comment by obackhouse** [comment]
> > What features, if any, are not supported by this PR? Does it support all self-consistencies, Fock loops etc.?
> 
> So I have tested everything seen in an example; evGW (neither, g0, w0), scGW (g0, w0, fock_loop and all combinations) and qsGW (eta, srg and solver (with combinations)). For chempot, see above and there is no UHF implemented. Is there anything else?

Sounds about everything - i guess we should try to make a nice `FEATURES.md` to list what combinations of features are actually available

**Review by obackhouse — APPROVED** [review]
LGTM - merge when ready

**Pull request merged**
